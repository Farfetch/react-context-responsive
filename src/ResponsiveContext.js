import { createContext } from 'react';

const ResponsiveContext = createContext({
    isCalculated: false,
    mediaType: 'xs',
    lessThan: {},
    greaterThan: {},
    is: {},
    orientation: null,
});

export default ResponsiveContext;
