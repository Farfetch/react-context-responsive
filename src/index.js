export { default as ResponsiveProvider } from './ResponsiveProvider';
export { default as useResponsive } from './useResponsive';
export { default as Responsive } from './Responsive';
export { default as withResponsive } from './withResponsive';
export { default as withIsMobile } from './withIsMobile';
export { default as useIsMobile } from './useIsMobile';
