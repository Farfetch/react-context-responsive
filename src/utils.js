export const getQueriesObjects = (currentBreakpointName, breakpoints) => {
    const currentBreakpoint = breakpoints.find(
        (breakpoint) => breakpoint.mediaType === currentBreakpointName
    );

    const defaultQueryObject = {
        is: {},
        greaterThan: {},
        lessThan: {},
    };

    if (!currentBreakpoint) {
        return defaultQueryObject;
    }

    return breakpoints.reduce((memo, breakpoint) => {
        const breakpointName = breakpoint.mediaType;
        return {
            is: {
                ...memo.is,
                [breakpointName]:
                    breakpointName === currentBreakpoint.mediaType,
            },
            greaterThan: {
                ...memo.greaterThan,
                [breakpointName]: currentBreakpoint.greaterThan.includes(
                    breakpointName
                ),
            },
            lessThan: {
                ...memo.lessThan,
                [breakpointName]: currentBreakpoint.lessThan.includes(
                    breakpointName
                ),
            },
        };
    }, defaultQueryObject);
};

export const getBreakpoints = (mediaQueries, breakpointNames) => {
    return breakpointNames.map((breakpointName, index) => {
        return {
            mediaType: breakpointName,
            mediaQuery: mediaQueries[breakpointName],
            greaterThan: breakpointNames.filter(
                (_gtName, gtIndex) => gtIndex < index
            ),
            lessThan: breakpointNames.filter(
                (_ltName, ltIndex) => ltIndex > index
            ),
        };
    }, {});
};

export const getMediaqueries = (breakpoints, breakpointsMax, breakpointNames) =>
    breakpointNames.reduce(
        (acc, breakpointName, index) => {
            const nextBreakpointName = breakpointNames[index + 1];

            return {
                ...acc,
                [breakpointName]: `(min-width: ${breakpoints[breakpointName]})${
                    nextBreakpointName
                        ? ` and (max-width: ${breakpointsMax[nextBreakpointName]})`
                        : ''
                }`,
            };
        },
        { _initial: '(min-width: 0em)' }
    );
