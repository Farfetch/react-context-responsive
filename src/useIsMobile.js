import { useContext, useMemo } from 'react';
import ResponsiveContext from './ResponsiveContext';

const useIsMobile = () => {
    const { isCalculated, lessThan } = useContext(ResponsiveContext);
    const isMobile = lessThan.md;

    return useMemo(() => ({ isMobile, isCalculated }), [
        isMobile,
        isCalculated,
    ]);
};

export default useIsMobile;
