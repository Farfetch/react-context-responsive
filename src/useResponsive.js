import { useContext } from 'react';
import ResponsiveContext from './ResponsiveContext';

const useResponsive = () => {
    const responsiveContext = useContext(ResponsiveContext);

    return responsiveContext;
};

export default useResponsive;
