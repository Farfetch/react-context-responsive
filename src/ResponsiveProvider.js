import { getBreakpoints, getMediaqueries, getQueriesObjects } from './utils';
import { useEffect } from 'react';
import PropTypes from 'prop-types';
import React, { useMemo, useRef, useState } from 'react';
import ResponsiveContext from './ResponsiveContext';

const ResponsiveProvider = ({
    initialMediaType,
    defaultOrientation,
    children,
    breakpoints,
    breakpointsMax,
    mediaQueries,
}) => {
    const [isCalculated, setIsCalculated] = useState(false);
    const [currentBreakpoint, setCurrentBreakpoint] = useState(
        initialMediaType
    );
    const [currentOrientation, setCurrentOrientation] = useState(
        defaultOrientation
    );

    const breakpointsWithInitialValue = {
        _initial: '0em',
        ...breakpoints,
    };

    const breakpointNames = Object.keys(
        mediaQueries || breakpointsWithInitialValue
    );

    if (!mediaQueries) {
        mediaQueries = getMediaqueries(
            breakpointsWithInitialValue,
            breakpointsMax,
            breakpointNames
        );
    }

    const breakpointsRef = useRef(
        getBreakpoints(mediaQueries, breakpointNames)
    );

    const queriesObjects = useMemo(
        () => getQueriesObjects(currentBreakpoint, breakpointsRef.current),
        [currentBreakpoint, breakpointsRef]
    );

    // Add event listener for breakpoints
    useEffect(() => {
        const breakpoints = breakpointsRef.current;
        const localMediaQueries = breakpoints.map(
            (breakpoint) => breakpoint.mediaQuery
        );
        const mediaQueryListsAndListeners = localMediaQueries.map(
            (mediaQuery, index) => {
                const mediaQueryList = window.matchMedia(mediaQuery);

                if (mediaQueryList.matches) {
                    setCurrentBreakpoint(breakpoints[index].mediaType);
                    setIsCalculated(true);
                }

                const listener = (event) => {
                    event.matches &&
                        setCurrentBreakpoint(breakpoints[index].mediaType);
                };

                mediaQueryList.addListener(listener);

                return [mediaQueryList, listener];
            }
        );

        return () => {
            mediaQueryListsAndListeners.forEach(
                ([mediaQueryList, listener]) => {
                    mediaQueryList.removeListener(listener);
                }
            );
        };
    }, [breakpointsRef]);

    // Add orientation media query
    useEffect(() => {
        const setMatchedOrientation = (matches) => {
            setCurrentOrientation(matches ? 'portrait' : 'landscape');
        };

        const orientationMediaQueryList = window.matchMedia(
            '(orientation: portrait)'
        );

        setMatchedOrientation(orientationMediaQueryList.matches);

        const listener = (event) => {
            setMatchedOrientation(event.matches);
        };

        orientationMediaQueryList.addListener(listener);

        return () => {
            orientationMediaQueryList.removeListener(listener);
        };
    }, []);

    const contextObject = useMemo(
        () => ({
            mediaType: currentBreakpoint,
            orientation: currentOrientation,
            isCalculated,
            ...queriesObjects,
        }),
        [currentBreakpoint, currentOrientation, isCalculated, queriesObjects]
    );

    useEffect(() => {
        if (process && process.env && process.env.NODE_ENV === 'development') {
            /* eslint-disable no-console */
            console.group(
                '%c @farfetch/react-context-responsive updated!',
                'color:#c4aa84; font-weight:bold;'
            );
            console.log(
                '%c Current breakpoint:',
                'font-weight:bold;',
                currentBreakpoint
            );
            console.log('%c Context:', 'font-weight:bold;', contextObject);
            console.groupEnd();
            /* eslint-enable no-console */
        }
    }, [contextObject, currentBreakpoint]);

    return (
        <ResponsiveContext.Provider value={contextObject}>
            {children}
        </ResponsiveContext.Provider>
    );
};

const breakpointsPropTypes = {
    xs: PropTypes.string,
    sm: PropTypes.string,
    md: PropTypes.string,
    lg: PropTypes.string,
    xl: PropTypes.string,
};

ResponsiveProvider.defaultProps = {
    initialMediaType: 'xs',
    defaultOrientation: null,
};

ResponsiveProvider.propTypes = {
    initialMediaType: PropTypes.oneOf([
        '_initial',
        'xs',
        'sm',
        'md',
        'lg',
        'xl',
    ]),
    defaultOrientation: PropTypes.oneOf(['landscape', 'portrait']),
    children: PropTypes.oneOfType([
        PropTypes.arrayOf(PropTypes.node),
        PropTypes.node,
    ]).isRequired,
    breakpoints: PropTypes.exact(breakpointsPropTypes),
    breakpointsMax: PropTypes.exact(breakpointsPropTypes),
    mediaQueries: PropTypes.exact({
        ...breakpointsPropTypes,
        _initial: PropTypes.string,
    }),
};

export default ResponsiveProvider;
