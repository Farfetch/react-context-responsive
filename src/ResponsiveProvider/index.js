import { getBreakpoints, getMediaqueries, getQueriesObjects } from '../utils';
import PropTypes from 'prop-types';
import React, { useMemo, useRef } from 'react';
import ResponsiveContext from '../ResponsiveContext';
import useDebugResponsive from './useDebugResponsive';
import useMediaType from './useMediaType';
import useOrientation from './useOrientation';

const ResponsiveProvider = ({
    initialMediaType,
    defaultOrientation,
    children,
    breakpoints,
    breakpointsMax,
    mediaQueries,
    mobileBreakpoint,
}) => {
    const breakpointsWithInitialValue = {
        _initial: '0em',
        ...breakpoints,
    };

    const breakpointNames = Object.keys(
        mediaQueries || breakpointsWithInitialValue
    );

    if (!mediaQueries) {
        mediaQueries = getMediaqueries(
            breakpointsWithInitialValue,
            breakpointsMax,
            breakpointNames
        );
    }

    const breakpointsRef = useRef(
        getBreakpoints(mediaQueries, breakpointNames)
    );

    const { currentMediaType, isCalculated } = useMediaType(
        breakpointsRef,
        initialMediaType
    );

    const queriesObjects = useMemo(
        () => getQueriesObjects(currentMediaType, breakpointsRef.current),
        [currentMediaType, breakpointsRef]
    );

    const currentOrientation = useOrientation(defaultOrientation);

    const contextObject = useMemo(
        () => ({
            mediaType: currentMediaType,
            orientation: currentOrientation,
            isCalculated,
            mobileBreakpoint,
            ...queriesObjects,
        }),
        [
            currentMediaType,
            currentOrientation,
            isCalculated,
            mobileBreakpoint,
            queriesObjects,
        ]
    );

    useDebugResponsive(contextObject, currentMediaType);

    return (
        <ResponsiveContext.Provider value={contextObject}>
            {children}
        </ResponsiveContext.Provider>
    );
};

const breakpointsPropTypes = {
    xs: PropTypes.string,
    sm: PropTypes.string,
    md: PropTypes.string,
    lg: PropTypes.string,
    xl: PropTypes.string,
};

ResponsiveProvider.defaultProps = {
    initialMediaType: 'xs',
    defaultOrientation: null,
    mobileBreakpoint: 'md',
};

ResponsiveProvider.propTypes = {
    initialMediaType: PropTypes.oneOf([
        '_initial',
        'xs',
        'sm',
        'md',
        'lg',
        'xl',
    ]),
    defaultOrientation: PropTypes.oneOf(['landscape', 'portrait']),
    children: PropTypes.oneOfType([
        PropTypes.arrayOf(PropTypes.node),
        PropTypes.node,
    ]).isRequired,
    breakpoints: PropTypes.exact(breakpointsPropTypes),
    breakpointsMax: PropTypes.exact(breakpointsPropTypes),
    mediaQueries: PropTypes.exact({
        ...breakpointsPropTypes,
        _initial: PropTypes.string,
    }),
    mobileBreakpoint: PropTypes.oneOf(['xs', 'sm', 'md', 'lg', 'xl']),
};

export default ResponsiveProvider;
