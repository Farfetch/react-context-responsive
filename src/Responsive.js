import PropTypes from 'prop-types';
import useResponsive from './useResponsive';

const Responsive = ({ children }) => {
    const responsive = useResponsive();

    return children(responsive);
};

Responsive.propTypes = {
    children: PropTypes.func.isRequired,
};

export default Responsive;
