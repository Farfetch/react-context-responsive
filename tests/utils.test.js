import { getBreakpoints, getQueriesObjects } from '../src/utils';
import mockBreakpoints from './__fixtures__/mockBreakpoints';

describe('getBreakpoints()', () => {
    test('should return the correct breakpoint list, in order', () => {
        expect(getBreakpoints()).toEqual(mockBreakpoints);
    });
});

const mediaTypes = mockBreakpoints.map((breakpoint) => breakpoint.mediaType);

describe('getQueriesObjects()', () => {
    test.each(mediaTypes)(
        'should return the correct query object for the %s breakpoint',
        (mediaType) => {
            expect(
                getQueriesObjects(mediaType, mockBreakpoints)
            ).toMatchSnapshot();
        }
    );
    test('should return all the fields with empty objects in case of an invalid breakpoint', () => {
        expect(getQueriesObjects('ab', mockBreakpoints)).toEqual({
            is: {},
            greaterThan: {},
            lessThan: {},
        });
    });
});
