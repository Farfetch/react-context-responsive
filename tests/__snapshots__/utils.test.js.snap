// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getQueriesObjects() should return the correct query object for the _initial breakpoint 1`] = `
Object {
  "greaterThan": Object {
    "_initial": false,
    "lg": false,
    "md": false,
    "sm": false,
    "xl": false,
    "xs": false,
  },
  "is": Object {
    "_initial": true,
    "lg": false,
    "md": false,
    "sm": false,
    "xl": false,
    "xs": false,
  },
  "lessThan": Object {
    "_initial": false,
    "lg": true,
    "md": true,
    "sm": true,
    "xl": true,
    "xs": true,
  },
}
`;

exports[`getQueriesObjects() should return the correct query object for the lg breakpoint 1`] = `
Object {
  "greaterThan": Object {
    "_initial": true,
    "lg": false,
    "md": true,
    "sm": true,
    "xl": false,
    "xs": true,
  },
  "is": Object {
    "_initial": false,
    "lg": true,
    "md": false,
    "sm": false,
    "xl": false,
    "xs": false,
  },
  "lessThan": Object {
    "_initial": false,
    "lg": false,
    "md": false,
    "sm": false,
    "xl": true,
    "xs": false,
  },
}
`;

exports[`getQueriesObjects() should return the correct query object for the md breakpoint 1`] = `
Object {
  "greaterThan": Object {
    "_initial": true,
    "lg": false,
    "md": false,
    "sm": true,
    "xl": false,
    "xs": true,
  },
  "is": Object {
    "_initial": false,
    "lg": false,
    "md": true,
    "sm": false,
    "xl": false,
    "xs": false,
  },
  "lessThan": Object {
    "_initial": false,
    "lg": true,
    "md": false,
    "sm": false,
    "xl": true,
    "xs": false,
  },
}
`;

exports[`getQueriesObjects() should return the correct query object for the sm breakpoint 1`] = `
Object {
  "greaterThan": Object {
    "_initial": true,
    "lg": false,
    "md": false,
    "sm": false,
    "xl": false,
    "xs": true,
  },
  "is": Object {
    "_initial": false,
    "lg": false,
    "md": false,
    "sm": true,
    "xl": false,
    "xs": false,
  },
  "lessThan": Object {
    "_initial": false,
    "lg": true,
    "md": true,
    "sm": false,
    "xl": true,
    "xs": false,
  },
}
`;

exports[`getQueriesObjects() should return the correct query object for the xl breakpoint 1`] = `
Object {
  "greaterThan": Object {
    "_initial": true,
    "lg": true,
    "md": true,
    "sm": true,
    "xl": false,
    "xs": true,
  },
  "is": Object {
    "_initial": false,
    "lg": false,
    "md": false,
    "sm": false,
    "xl": true,
    "xs": false,
  },
  "lessThan": Object {
    "_initial": false,
    "lg": false,
    "md": false,
    "sm": false,
    "xl": false,
    "xs": false,
  },
}
`;

exports[`getQueriesObjects() should return the correct query object for the xs breakpoint 1`] = `
Object {
  "greaterThan": Object {
    "_initial": true,
    "lg": false,
    "md": false,
    "sm": false,
    "xl": false,
    "xs": false,
  },
  "is": Object {
    "_initial": false,
    "lg": false,
    "md": false,
    "sm": false,
    "xl": false,
    "xs": true,
  },
  "lessThan": Object {
    "_initial": false,
    "lg": true,
    "md": true,
    "sm": true,
    "xl": true,
    "xs": false,
  },
}
`;
